substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		// closure to calculate the stake imbalance of a staker.
+		let stake_imbalance_of = |prev_stake: Option<Stake<BalanceOf<T>>>,
+		                          voter_weight: ExtendedBalance| {
+			if let Some(prev_stake) = prev_stake {
+				let prev_voter_weight = Self::to_vote_extended(prev_stake.active);
+
+				if prev_voter_weight > voter_weight {
+					StakeImbalance::Negative(
+						prev_voter_weight.defensive_saturating_sub(voter_weight),
+					)
+				} else {
+					StakeImbalance::Positive(
+						voter_weight.defensive_saturating_sub(prev_voter_weight),
+					)
+				}
+			} else {
+				// if staker had no stake before update, then add all the voter weight
+				// to the target's score.
+				StakeImbalance::Positive(voter_weight)
+			}
+		};
+
+		if T::Staking::status(who)
+			.and(T::Staking::stake(who))
+			.defensive_proof(
+				"staker should exist when calling on_stake_update and have a valid status",
+			)
+			.is_ok()
+		{
+			let voter_weight = Self::weight_of(stake.active);
+
+			match T::Staking::status(who).expect("status checked above; qed.") {
+				StakerStatus::Nominator(nominations) => {
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"staker should exist in VoterList, as per the contract \
+                            with staking.",
+					);
+
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+
+					log!(
+						debug,
+						"on_stake_update: {:?} with {:?}. impacting nominations {:?}",
+						who,
+						stake_imbalance,
+						nominations,
+					);
+
+					// updates vote weight of nominated targets accordingly. Note: this will
+					// update the score of up to `T::MaxNominations` validators.
+					for target in nominations.into_iter() {
+						Self::update_target_score(&target, stake_imbalance);
+					}
+				},
+				StakerStatus::Validator => {
+					// validator is both a target and a voter.
+					let stake_imbalance = stake_imbalance_of(prev_stake, voter_weight.into());
+
+					Self::update_target_score(who, stake_imbalance);
+
+					let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+						"the staker should exit in VoterList, as per the \
+                            contract with staking.",
+					);
+				},
+				StakerStatus::Idle => (), // nothing to see here.
+			}
+		}
+	}
+
+	/// Fired when someone sets their intention to validate.
+	///
+	/// A validator is also considered a voter with self-vote and should also be added to
+	/// [`Config::VoterList`].
+	//
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* calling this
+	/// method.
+	fn on_validator_add(who: &T::AccountId, self_stake: Option<Stake<BalanceOf<T>>>) {
+		let self_stake = self_stake.unwrap_or_default().active;
+
+		if !T::TargetList::contains(who) {
+			T::TargetList::on_insert(who.clone(), self_stake)
```

Comment:
`on_insert` docs say `implementation should return an error if duplicate item is being inserted`, so you can get rid of the extra access in `T::TargetList::contains` by just handling the error of `T::TargetList::on_insert`.

Reply:
Thanks, I simplified the code.

Reply:
> You can also exit the function early for StakingStatus::Idle and do the voter list update sooner, outside of the specialized logic for the target list.

Thanks, I simplified the code.

>  so you can get rid of the extra access in T::TargetList::contains by just handling the error of T::TargetList::on_insert

I agree, removed the extra `contains`, thanks!